gh-101100: Fix Sphinx warnings in library/random.rst (#112981)
python/cpython: Doc/library/random.rst

instances of :class:`Random` to get generators that don't share state.

Class :class:`Random` can also be subclassed if you want to use a different
basic generator of your own devising: in that case, override the :meth:`~random.random`,
:meth:`~random.seed`, :meth:`~random.getstate`, and :meth:`~random.setstate` methods.

> Optionally, a new generator can supply a :meth:`~random.getrandbits` method --- this

Reviewer: I am not sure that it is right to replace references to Random methods by references to global functions.

@rhettinger, what do you think?

Reviewer 2: @serhiy-storchaka I also think the proposed edit is incorrect.

Author: Okay, how's this? https://github.com/python/cpython/pull/112981/commits/4bbac89bdb87d2db0b715b93fe2341aae9258e9b

Or should they be separately documented?

   .. deprecated-removed:: 3.9 3.11
      Formerly the *seed* could be any hashable object.  Now it is limited to:

>       *NoneType*, :class:`int`, :class:`float`, :class:`str`,

Reviewer: Why is this change? The right reference is ```:class:`~types.NoneType` ```, but usually we just use ``` ``None`` ```.

Author: To match how it was referred to elsewhere in this file:

https://github.com/python/cpython/blob/d3d310fa6938e99c98dd900f7310f1b202519d73/Doc/library/random.rst?plain=1#L89-L92

Shall I change them both? To which version?

Reviewer: Please change it to ``` ``None`` ```. `NoneType` only occurs in the output of examples and in the `None` and `NoneType` documentation. Except random.rst.

Author: Updated!

      ``None``, :class:`int`, :class:`float`, :class:`str`,
      :class:`bytes`, or :class:`bytearray`.

   Subclasses of :class:`!Random` should override the following methods if they
   wish to make use of a different basic generator:

   .. method:: Random.seed(a=None, version=2)

>       Override this method in subclasses to customise the :meth:`random.seed`
>       behaviour of :class:`Random` instances.
> 
>    .. method:: Random.getstate()
> 
>       Override this method in subclasses to customise the :meth:`random.getstate`
>       behaviour of :class:`Random` instances.
> 
>    .. method:: Random.setstate(state)
> 
>       Override this method in subclasses to customise the :meth:`random.setstate`
>       behaviour of :class:`Random` instances.
> 
>    .. method:: Random.random()
> 
>       Override this method in subclasses to customise the :meth:`random.random`
>       behaviour of :class:`Random` instances.
> 
>    Optionally, a new generator can also supply the following method:
> 
>    .. method:: Random.getrandbits(k)
> 
>       Override this method in subclasses to customise the
>       :meth:`random.getrandbits` behaviour of :class:`Random` instances. This in

Reviewer 3: Some squigglies might make these slightly more concise to read:

```suggestion
      Override this method in subclasses to customise the :meth:`~random.seed`
      behaviour of :class:`Random` instances.

   .. method:: Random.getstate()

      Override this method in subclasses to customise the :meth:`~random.getstate`
      behaviour of :class:`Random` instances.

   .. method:: Random.setstate(state)

      Override this method in subclasses to customise the :meth:`~random.setstate`
      behaviour of :class:`Random` instances.

   .. method:: Random.random()

      Override this method in subclasses to customise the :meth:`~random.random`
      behaviour of :class:`Random` instances.

   Optionally, a new generator can also supply the following method:

   .. method:: Random.getrandbits(k)

      Override this method in subclasses to customise the
      :meth:`~random.getrandbits` behaviour of :class:`Random` instances. This in
```

>    Optionally, a new generator can also supply the following method:
> 
>    .. method:: Random.getrandbits(k)
> 
>       Override this method in subclasses to customise the
>       :meth:`~random.getrandbits` behaviour of :class:`!Random` instances. This in
>       turn allows :meth:`!Random.randrange` to produce selections over an
>       arbitrarily large range.

Reviewer 3: Actually, this might be slightly better wording; not sure about my previous suggestion:

```suggestion
   Optionally, a custom generator subclass can also supply the following method:

   .. method:: Random.getrandbits(k)

      Override this method in subclasses to customise the
      :meth:`~random.getrandbits` behaviour of :class:`!Random` instances.
      (:meth:`!getrandbits` is used by
      :meth:`Random.randrange <random.randrange>` to produce selections over an
      arbitrarily large range.)
```

WDYT?

Author: Fine by me, or like the others, we could avoid duplicating stuff from the linked `random.getrandbits()`?

```suggestion
   Optionally, a new generator can also supply the following method:

   .. method:: Random.getrandbits(k)

      Override this method in subclasses to customise the
      :meth:`~random.getrandbits` behaviour of :class:`!Random` instances.
```

Reviewer 3: Ah good point, I hadn't spotted that the docs for `getrandbits` also covered that! A small tweak to your suggestion:

```suggestion
   Optionally, a custom generator subclass can also supply the following method:

   .. method:: Random.getrandbits(k)

      Override this method in subclasses to customise the
      :meth:`~random.getrandbits` behaviour of :class:`!Random` instances.
```

      ``None``, :class:`int`, :class:`float`, :class:`str`,
      :class:`bytes`, or :class:`bytearray`.

>    .. method:: Random.seed()

Reviewer 3: Is this signature correct? The docs for `random.seed()` give `random.seed(a=None, version=2)` https://docs.python.org/3/library/random.html#random.seed

      Override this method in subclasses to customise the :meth:`random.seed`
      behaviour of :class:`Random` instances.

   .. method:: Random.getstate()

      Override this method in subclasses to customise the :meth:`random.getstate`
      behaviour of :class:`Random` instances.

>    .. method:: Random.setstate()

Reviewer 3: Also not sure about this signature: https://docs.python.org/3/library/random.html#random.setstate

      Override this method in subclasses to customise the :meth:`random.setstate`
      behaviour of :class:`Random` instances.

>    .. method:: Random.getrandbits()

Reviewer 3: Or this one: https://docs.python.org/3/library/random.html#random.setstate

   including simulation, sampling, shuffling, and cross-validation.

   `Economics Simulation

>    <https://nbviewer.org/url/norvig.com/ipython/Economics.ipynb>`_

Reviewer 4: @hugovk I believe that `nbviewer.jupyter.org` is the correct URL. Thanks. Same as the other link.

Author: When I click the `nbviewer.jupyter.org` links I'm redirected to `nbviewer.org`, is it different for you? We can keep the originals if you prefer, they do still work.

Reviewer: https://nbviewer.jupyter.org/url/norvig.com/ipython/Economics.ipynb returns HTTP/2 302 and redirects to https://nbviewer.org/url/norvig.com/ipython/Economics.ipynb. While the current link works, the new link looks more canonical.

Reviewer 4: Yeah, I get the redirect too. If I use only `nbviewer.org`, I was getting 500 errors but I suppose that may be due to Cloudflare's WARP on my machine. The URL that you updated seems to work in incognito mode so let's leave as you changed. Thanks!
